src/garage/_environment.py

```diff
@@ -0,0 +1,226 @@
+    """The main API for garage environments.
+
+    The public API methods are:
+
+    +-----------------------+
+    | Functions             |
+    +=======================+
+    | reset()               |
+    +-----------------------+
+    | step()                |
+    +-----------------------+
+    | render()              |
+    +-----------------------+
+    | visualize()           |
+    +-----------------------+
+    | close()               |
+    +-----------------------+
+
+    Set the following properties:
+
+    +-----------------------+-------------------------------------------------+
+    | Properties            | Description                                     |
+    +=======================+=================================================+
+    | action_space          | The action space specification                  |
+    +-----------------------+-------------------------------------------------+
+    | observation_space     | The observation space specification             |
+    +-----------------------+-------------------------------------------------+
+    | spec                  | The environment specifications                  |
+    +-----------------------+-------------------------------------------------+
+    | render_modes          | The list of supported render modes              |
+    +-----------------------+-------------------------------------------------+
+
+    Example of a simple rollout loop:
+
+    .. code-block:: python
+
+        env = MyEnv()
+        policy = MyPolicy()
+        first_ts = env.reset()
+        env.visualize()  # visualization window opened
+
+        episode = [first_ts, ]
+        while not ts.last():
+            action = policy.get_action(episode[-1].next_observation)
+            episode.append(env.step(action))
+
+        env.close()  # visualization window closed
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__setstate__()` and
+        `__getstate()__` to add your custom pickle logic.
+
+        You might want to refer to the EzPickle module:
+        https://github.com/openai/gym/blob/master/gym/utils/ezpickle.py
+        for a lightweight way of pickle and unpickle via constructor
+        arguments.
+
+    """
+
+    @property
+    @abc.abstractmethod
+    def action_space(self):
+        """Return the action space specification.
+
+        Some environments do not provide their action space specifications as
+        type `akro.Space`. In such cases, subclasses should silently convert
+        the environment's action_space to `akro.Space`.
+
+        Returns:
+            akro.Space: The environment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def observation_space(self):
```

Comment:
see my previous comment about how to document properties (i.e. as if they are attributes)